src/rpc/blockchain.cpp

```diff
@@ -518,6 +518,69 @@ UniValue getblockbynumber(const UniValue& params, bool fHelp)
     return blockToJSON(block, pblockindex, params.size() > 1 ? params[1].get_bool() : false);
 }
 
+UniValue getblocksbatch(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() < 1 || params.size() > 3)
+        throw runtime_error(
+                "getblocksbatch <starting number> <number of blocks> [bool:txinfo]\n"
+                "\n"
+                "<starting number> the block number at the start of the batch"
+                "\n"
+                "<number of blocks> the number of blocks to return in the batch, limited to 500"
+                "\n"
+                "[bool:txinfo] optional to print more detailed tx info\n"
+                "\n"
+                "Returns a JSON array with details of the requested blocks starting with\n"
+                "the given block-number.\n");
+
+    UniValue result(UniValue::VOBJ);
+    UniValue blocks(UniValue::VARR);
+
+    int nHeight = params[0].get_int();
+    if (nHeight < 0 || nHeight > nBestHeight) throw runtime_error("Starting block number out of range");
+
+    int batch_size = params[1].get_int();
+    if (batch_size < 1 || batch_size > 500)
+    {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Batch size must be between 1 and 500, inclusive.");
+    }
+
+    bool transaction_details = false;
+    if (params.size() > 2) transaction_details = params[2].get_bool();
+
+    LOCK(cs_main);
+
+    // Select the block index for the head of the chain.
+    CBlockIndex* pblockindex_head = mapBlockIndex[hashBestChain];
+    CBlockIndex* pblockindex = pblockindex_head;
+
+    // Rewind to the block corresponding to the specified height.
+    while (pblockindex->nHeight > nHeight)
+    {
+        pblockindex = pblockindex->pprev;
+    }
+
+    int i = 0;
+    while (i < batch_size)
+    {
+        CBlock block;
+        if (!block.ReadFromDisk(pblockindex, true)) throw runtime_error("Error reading block from specified batch.");
+
+        blocks.push_back(blockToJSON(block, pblockindex, transaction_details));
+        ++i;
+
+        if (pblockindex == pblockindex_head) break;
+
+        pblockindex = pblockindex->pnext;
+    }
+
+    result.pushKV("starting_block_number", nHeight);
+    result.pushKV("block_count", i);
```

Comment:
Are passing these back necessary?

Reply:
Yes. In situations where the starting block is less than batch_size away from the head, a smaller number of blocks will be returned than the batch size. This is provided for convenience as this will likely be called in a loop exterior to the wallet, and the "last" batch in the loop is likely to be of this condition. It the calling code doesn't need them, it doesn't have to use them.